refactor(compiler): account for pipes without names

crisbeto opened the pull request:
Updates the compiler logic to account for pipe definitions that may not have names.

devversion commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/pipes/GOLDEN_PARTIAL.js:
```diff
@@ -243,3 +243,30 @@ export declare class MyModule {
+export declare class PipeWithoutName implements PipeTransform {
+    transform(value: unknown): unknown;
+    static ɵfac: i0.ɵɵFactoryDeclaration<PipeWithoutName, never>;
+    static ɵpipe: i0.ɵɵPipeDeclaration<PipeWithoutName, null, true>;
```
I'm curious, do you have thoughts this output appearing in projects using an older compiler version? I don't fully recall what our version guarantees here are (which way, backwards, forward etc.).

Likely simply not picking up the pipe metadata is the expected behavior here.

crisbeto replied:
Hmm not entirely sure either, but I suspect that we don't support this sort of backwards compatibility.

devversion replied:
Yeah, it's probably fine because there is no way to reference the pipe nicely "without" a name; but might be worth double-checking with others.

Worst-case, we could roll out changes to older compilers and error if we discover a nameless pipe when collecting metadata for a symbol via `.d.ts`

JoostK replied:
Reporting an error when used with older compilers can be done by changing the `minVersion` value of the partial declaration, that's what it's designed for.

devversion commented on packages/compiler-cli/test/compliance/test_cases/r3_compiler_compliance/components_and_directives/pipes/GOLDEN_PARTIAL.js:
```diff
@@ -243,3 +243,30 @@ export declare class MyModule {
+    }
+}
+PipeWithoutName.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "0.0.0-PLACEHOLDER", ngImport: i0, type: PipeWithoutName, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
+PipeWithoutName.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "0.0.0-PLACEHOLDER", ngImport: i0, type: PipeWithoutName, isStandalone: true, name: "PipeWithoutName" });
```
Out of curiosity, you did choose the `name` to be still emitted for backwards compatibility of such "nameless pipe" partial compilation output, right?

crisbeto replied:
Yeah I left it for backwards compatibility and because this is the name that the runtime will use to "match" pipes.

devversion reviewed (approved):
LGTM

Reviewed-for: public-api

crisbeto commented:
[Passing TGP](http://test/OCL:758613863:BASE:758715729:1747240171859:d59c980e)

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-general

crisbeto commented:
There were some discussions over whether we want to do the public API change. I've dropped it for now and will leave the rest of the compiler code so that we can easily make the switch later.

crisbeto committed:
refactor(compiler): account for pipes without names

Updates the compiler logic to account for pipe definitions that may not have names.

thePunderWoman commented:
This PR was merged into the repository by commit 7fa3203ee763756919b7d03ff19b903a27a5e1d6.

The changes were merged into the following branches: main
